tests/TestCase/Collection/CollectionTest.php

```diff
@@ -1163,4 +1163,46 @@ public function testThroughReturnArray()
 
         $this->assertEquals([1, 2, 3, 1, 2, 3], $collection->toList());
     }
+
+    /**
+     * Tests __debugInfo() or debug() usage
+     *
+     * @return void
+     */
+    public function testDebug()
+    {
+        $items = [1, 2, 3];
+
+        $collection = new Collection($items);
+
+        $result = $collection->__debugInfo();
+        $expected = [
+            'count' => 3,
+        ];
+        $this->assertSame($expected, $result);
+
+        // Calling it again will rewind
+        $result = $collection->__debugInfo();
+        $expected = [
+            'count' => 3,
+        ];
+        $this->assertSame($expected, $result);
+
+        // Make sure it also works with non rewindable iterators
+        $iterator = new NoRewindIterator(new ArrayIterator($items));
+        $collection = new Collection($iterator);
+
+        $result = $collection->__debugInfo();
+        $expected = [
+            'count' => 3,
+        ];
+        $this->assertSame($expected, $result);
+
+        // Calling it again will in this case not rewind
+        $result = $collection->__debugInfo();
+        $expected = [
+            'count' =>  0,
+        ];
+        $this->assertSame($expected, $result);
```

Comment:
Since it cannot rewind, those will be vanished now. But those things are to be expected in that case IMO.
It it like `debug($this->Auth->login()); $this->Auth->login();` in 2.x as many people used and expected the second method call to be the same as the first, whereas it would then return false because the first one returned true.
More appropriate would have been `debug($this->Auth->login()); die(); $this->Auth->login();`.

So when debugging this kind of iterator script execution needs to be stopped until the debug() call has been removed again for it to make sense.
This would be the case with any operation on it.